File: integration-kits/python/cdk_spilman_kit/configurable.py
import os
import yaml
import json
from typing import Dict, List, Any, Optional
from cdk_spilman import SpilmanBridge
from .host import BaseSpilmanHost
from .stores import SpilmanStores, SqliteSpilmanStores
from .keysets import refresh_keyset_cache

class ConfigurableSpilman:
    def __init__(
        self, 
        config: Dict[str, Any], 
        stores: SpilmanStores, 
        host: BaseSpilmanHost, 
        bridge: SpilmanBridge,
        spilman: Optional[Any] = None
    ):
        self.config = config
        self.stores = stores
        self.host = host
        self.bridge = bridge
        self.spilman = spilman

    @staticmethod
    def from_yaml(config_path: str, secret_key_hex: str) -> "ConfigurableSpilman":
        with open(config_path, "r") as f:
            config = yaml.safe_load(f)

        # Support MINT_URL override like Rust/TS
        if "MINT_URL" in os.environ:
            mint_url = os.environ["MINT_URL"]
            all_units = list(config.get("pricing", {}).keys())
            config["mints"] = {mint_url: all_units}

        pricing = config.get("pricing", {})

        pricing_scale = config.get("pricing_scale", 1)
        try:
            pricing_scale = int(pricing_scale)
        except (TypeError, ValueError):
            pricing_scale = 1
        if pricing_scale <= 0:
            pricing_scale = 1
        config["pricing_scale"] = pricing_scale

        # Initialize stores
        storage_cfg = config.get("storage", {})
        if storage_cfg.get("type") == "sqlite" and storage_cfg.get("path"):
            stores = SqliteSpilmanStores(storage_cfg["path"])
        else:
            stores = SpilmanStores()

        host = BaseSpilmanHost(
            secret_key=secret_key_hex,
            mints=config.get("mints", {}),
            pricing=pricing,
            stores=stores,
            min_expiry_seconds=config.get("min_expiry_seconds", 3600),
            pricing_scale=pricing_scale,
        )

        bridge = SpilmanBridge(host)
        
        instance = ConfigurableSpilman(config, stores, host, bridge)
        
        # Initial keyset refresh
        instance.initialize_keysets()
        
        return instance

    def initialize_keysets(self):
        for mint_url in self.config.get("mints", {}).keys():
            try:
                self.host.refresh_all_keysets(mint_url)
            except Exception as e:
                print(f"WARNING: Failed to fetch keysets from {mint_url}: {e}")

    def init_flask(self, app):
        from .ext.flask import Spilman
        self.spilman = Spilman(app, self.host, self.bridge)
        return self.spilman

File: integration-kits/go/configurable.go
package spilmankit

import (
	"os"

	"github.com/cashubtc/spilman-go/spilman"
	"gopkg.in/yaml.v3"
)

type SpilmanConfig struct {
	Mints            map[string][]string `yaml:"mints"`
	MinExpirySeconds uint64              `yaml:"min_expiry_seconds"`
	PricingScale     uint64              `yaml:"pricing_scale"`
	Storage          struct {
		Type string `yaml:"type"`
		Path string `yaml:"path"`
	} `yaml:"storage"`
	Pricing PricingTable `yaml:"pricing"`
}

type ConfigurableSpilman struct {
	Config *SpilmanConfig
	Stores SpilmanStores
	Host   *BaseSpilmanHost
	Bridge *spilman.Bridge
}

func LoadFromYaml(configPath, secretKeyHex string) (*ConfigurableSpilman, error) {
	yamlFile, err := os.ReadFile(configPath)
	if err != nil {
		return nil, err
	}

	var config SpilmanConfig
	err = yaml.Unmarshal(yamlFile, &config)
	if err != nil {
		return nil, err
	}

	// Support MINT_URL override like Rust/TS/Python
	if mintUrl := os.Getenv("MINT_URL"); mintUrl != "" {
		allUnits := []string{}
		for u := range config.Pricing {
			allUnits = append(allUnits, u)
		}
		config.Mints = map[string][]string{mintUrl: allUnits}
	}

	var stores SpilmanStores
	if config.Storage.Type == "sqlite" && config.Storage.Path != "" {
		stores, err = NewSqliteStores(config.Storage.Path)
		if err != nil {
			return nil, err
		}
	} else {
		stores = NewInMemoryStores()
	}

	if config.MinExpirySeconds == 0 {
		config.MinExpirySeconds = 3600
	}
	if config.PricingScale == 0 {
		config.PricingScale = 1
	}

	host := NewBaseSpilmanHost(secretKeyHex, config.Mints, config.Pricing, stores, config.MinExpirySeconds, config.PricingScale)
	bridge := spilman.NewBridge(host)

	ctx := &ConfigurableSpilman{
		Config: &config,
		Stores: stores,
		Host:   host,
		Bridge: bridge,
	}

	// Initial keyset refresh (non-blocking)
	go ctx.InitializeKeysets()

	return ctx, nil
}

func (c *ConfigurableSpilman) InitializeKeysets() {
	for mintUrl := range c.Config.Mints {
		c.Host.RefreshAllKeysets(mintUrl)
	}
}

func (c *ConfigurableSpilman) Free() {
	if c.Bridge != nil {
		c.Bridge.Free()
	}
}
